### PR TITLE
RFC: online schema change.

### PR DESCRIPTION
**Pull request by petermattis**
[ci skip]

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/2842)

### CODE REVIEW

**Comment by petermattis** [comment]
Sending out for initial comments from our European contingent, though I haven't spent enough time on this yet.

File: docs/RFCS/online_schema_change.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+changes on a table (e.g. concurrently adding multiple indexes).
+
+# Drawbacks
+
+# Alternatives
+
+# Unresolved questions
+
+* Is there a way to avoid performing the backfilling work
+  transactionally? Is it worth optimizing?
```
Comment (tschottdorf) [inline docs/RFCS/online_schema_change.md]: It's definitely going to be worth optimizing. Possibly we can avoid the transaction if we are operating under the assumption that, for example, the raw index kvs we're backfilling started out empty before the schema change. Then any key with a recent write timestamp is already up to date and need not be touched. But that's pretty borderline terrain and is self-contained enough to worry about in the future.

Comment (tschottdorf) [inline docs/RFCS/online_schema_change.md]: A more obvious optimization is persisting which key ranges have already been backfilled, so that an interrupted backfill avoids duplicate work.

File: docs/RFCS/online_schema_change.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    })
+    if err != nil {
+      // Abort!
+    }
+    startKey = lastKey.Next()
+  }
+```
+
+The above pseudo-code is intended to give the gist of how backfilling
+will work. We'll iterate over the primary indexn transactionally
```
Comment (tschottdorf) [inline docs/RFCS/online_schema_change.md]: Typo

File: docs/RFCS/online_schema_change.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+Since schema change operations are potentially long running they need
+to be restartable or abortable if the node performing them dies. We
+accomplish this by performing the schema change operation for a table
+on a well known node: the replica holding the leader lease for the
+first range of the table (i.e. containing the key `/<tableID>`). When
+a node receives a schema change operation such as `CREATE INDEX` it
+will forward the operation to this "table leader". When the table
+leader restarts it will load the associated table descriptor and
+restart or abort the schema change operation.
```
Comment (tschottdorf) [inline docs/RFCS/online_schema_change.md]: An abort is still going to be a forward schema change. Can't hurt to mention that.

**Comment by tschottdorf** [comment]
LGTM

File: docs/RFCS/online_schema_change.md (hunk 2 above)
Comment (petermattis) [inline docs/RFCS/online_schema_change.md]: Fixed.

File: docs/RFCS/online_schema_change.md (hunk 1 above)
Comment (petermattis) [inline docs/RFCS/online_schema_change.md]: Good point about restarting interrupted backfills. I'll add a note about that.

Agreed that we could look at the write timestamp to determine if the index key needs to be written. That seems expensive with the existing KV operations. What do you think of adding a `PutIfOlderThan(key, value, timetamp)` operation? Or perhaps just make `OlderThan` an option on regular `Put` operations? Then we could grab the timestamp before starting the index creation (e.g. before the index is added to the descriptor in delete-only mode). Then any write operations that are (including deletions) that are newer than that timestamp should be preferred over the index backfill.

File: docs/RFCS/online_schema_change.md (hunk 3 above)
Comment (petermattis) [inline docs/RFCS/online_schema_change.md]: By "forward schema change" do you mean it will cause the descriptor version to increase?

Comment (tschottdorf) [inline docs/RFCS/online_schema_change.md]: Yes, at least my impression was that you never just get to 'delete' a newer
descriptor but instead you add a new one on top. If that's not the plan, an
outline for aborting would be appreciated.

**Comment by petermattis** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

<sup>**[docs/RFCS/online_schema_change.md, line 129 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842#-K0fgd-jqUeAyY4hgQNQ-r1-129)** ([raw file](https://github.com/cockroachdb/cockroach/blob/fb5d6c1d13ecb075a0ab4646bc5ca69edb059206/docs/RFCS/online_schema_change.md#L129)):</sup>
Yes, that's correct. You never rollback to a previous descriptor versions but instead perform a compensating mutation to the descriptor to create a new version. I've made this slightly clearer in the doc.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842)

File: docs/RFCS/online_schema_change.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+  oneof mutation {
+    optional IndexDescriptor add_index;
+    optional IndexDescriptor drop_index;
+    optional ColumnDescriptor add_column;
+    optional ColumnDescriptor drop_column;
+  }
+}
+```
+
+Additionally, index and column descriptors will be enhanced to contain
```
Comment (vivekmenezes) [inline docs/RFCS/online_schema_change.md]: Besides what we have below to preserve the state of a mutation, is there a need for the mutation oneof above.

**Comment by petermattis** [comment]
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

<sup>**[docs/RFCS/online_schema_change.md, line 51 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842#-K0fgd-jqUeAyY4hgQNQ-r3-51)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0574252634553b297160bc080ca5232f3af4e1/docs/RFCS/online_schema_change.md#L51)):</sup>
I'm not entirely sure. The thought was the `oneof mutation` makes it obvious what sort of schema change is being performed. Simply relying on the `State` below might leave some ambiguity as to whether an index is being added or removed.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842)

File: docs/RFCS/online_schema_change.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+  for startKey != endKey {
+    var lastKey roachpb.Key
+    err := db.Txn(func(txn *Txn) error {
+      txn.SetPriority(VeryLowPriority)
+      scan, err := txn.Scan(startKey, endKey, 1000)
+      if err != nil {
+        return err
+      }
+      lastKey = getLastKeyOfFullRowInScan(scan)
+      b := makeIndexKeysBatch(scan)
```
Comment (vivekmenezes) [inline docs/RFCS/online_schema_change.md]: It's worth adding a checkpoint somewhere every once in a while. perhaps we should create a separate table for checkpoints.

File: docs/RFCS/online_schema_change.md [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+* No real good ones. We could try to come up with some sort of global
+  table lock mechanism and then synchronously perform the
+  backfill. This seems as difficult as the current proposal and much
+  worse for the user experience.
+
+# Unresolved questions
+
+* Is there a way to avoid performing the backfilling work
+  transactionally? Is it worth optimizing? Delete, in particular,
```
Comment (vivekmenezes) [inline docs/RFCS/online_schema_change.md]: no

File: docs/RFCS/online_schema_change.md [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+  seems problematic. If we scan a portion of the primary index
+  non-transactionally and then generate index keys, a concurrent
+  delete can come in and delete one of the rows and not generate a
+  delete of the index key unless we left a "tombstone" deletion for
+  the index key.
+
+* If the node performing the backfill gets restarted we should figure
+  out a way to avoid restarting the backfill from scratch. One thought
+  is that the backfill operation can periodically checkpoint the high
+  water mark of its progress (perhaps in the descriptor itself).
```
Comment (vivekmenezes) [inline docs/RFCS/online_schema_change.md]: you do talk about checkpointing here. We could add it to the descriptor but we are not allowed to change the version. I'd add it as another column to the descriptor table

File: docs/RFCS/online_schema_change.md [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+* Figure out how to distribute the backfill work. Ideally we would
+  have each range of the primary index generate and write the index
+  keys. I feel this is best left to a separate RFC.
```
Comment (vivekmenezes) [inline docs/RFCS/online_schema_change.md]: We should do this when we have a distributed sql implementation.

**Comment by vivekmenezes** [comment]
LGTM

**Comment by petermattis** [comment]
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

<sup>**[docs/RFCS/online_schema_change.md, line 103 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842#-K0fgd-jqUeAyY4hgQNQ-r3-103)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0574252634553b297160bc080ca5232f3af4e1/docs/RFCS/online_schema_change.md#L103)):</sup>
There is a note about this in the `Unresolved Questions` section. 


---

<sup>**[docs/RFCS/online_schema_change.md, line 152 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842#-K0fgd-jqUeAyY4hgQNQ-r3-152)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0574252634553b297160bc080ca5232f3af4e1/docs/RFCS/online_schema_change.md#L152)):</sup>
Eh? Do you have a proof that it has to be done transactionally? I certainly agree that the result has to leave the table data in a consistent state, but that doesn't necessary lead to usage of a transaction. My comment to @tschottdorf outlines how we could do it non-transactionally. 


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842)

**Comment by petermattis** [comment]
Review status: 0 of 1 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

<sup>**[docs/RFCS/online_schema_change.md, line 162 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842#-K0fgd-jqUeAyY4hgQNQ-r3-162)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0574252634553b297160bc080ca5232f3af4e1/docs/RFCS/online_schema_change.md#L162)):</sup>
Good point about not changing the descriptor version, though that doesn't mean we can't store a bit of data in the descriptor (we can overwrite the existing descriptor without changing the version as long as the serving state of the descriptor is unchanged). Not sure if it matters too much where the checkpoint is stored.


---

<sup>**[docs/RFCS/online_schema_change.md, line 166 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842#-K0fgd-jqUeAyY4hgQNQ-r3-166)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0574252634553b297160bc080ca5232f3af4e1/docs/RFCS/online_schema_change.md#L166)):</sup>
I wasn't proposing to do this work now. Certainly isn't necessary in the short term.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842)

File: docs/RFCS/online_schema_change.md (hunk 1 above)
Comment (tschottdorf) [inline docs/RFCS/online_schema_change.md]: Yes, something like this should work.

**Comment by bdarnell** [comment]
LGTM


---

Reviewed 1 of 1 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

<sup>**[docs/RFCS/online_schema_change.md, line 70 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842#-K0fwU017HH9MP6eNFNl)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0574252634553b297160bc080ca5232f3af4e1/docs/RFCS/online_schema_change.md#L70)):</sup>
We should point to the F1 paper's diagram of the state transitions (page 9 of http://static.googleusercontent.com/media/research.google.com/en//pubs/archive/41376.pdf) or make our own.


---

<sup>**[docs/RFCS/online_schema_change.md, line 83 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842#-K0fuU-9jkSmRqSyowhV)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0574252634553b297160bc080ca5232f3af4e1/docs/RFCS/online_schema_change.md#L83)):</sup>
Is there a reverse backfill (back-empty?) process when the index is in the `DELETE_ONLY` state? Or are the old index records removed en masse after the index has been transitioned from `DELETE_ONLY` to non-existent?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842)

**Comment by petermattis** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

<sup>**[docs/RFCS/online_schema_change.md, line 70 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842#-K0fwU017HH9MP6eNFNl)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0574252634553b297160bc080ca5232f3af4e1/docs/RFCS/online_schema_change.md#L70)):</sup>
I'll point to the F1 paper for now.


---

<sup>**[docs/RFCS/online_schema_change.md, line 83 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842#-K0fuU-9jkSmRqSyowhV)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0574252634553b297160bc080ca5232f3af4e1/docs/RFCS/online_schema_change.md#L83)):</sup>
Back-delete? Deleting index data is straightforward as we can issue a single `DelRange` op. Deleting column data is harder and will require something like a back-delete operation. 


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842)

**Comment by vivekmenezes** [comment]
Review status: 0 of 1 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

<sup>**[docs/RFCS/online_schema_change.md, line 51 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842#-K0fgd-jqUeAyY4hgQNQ-r3-51)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0574252634553b297160bc080ca5232f3af4e1/docs/RFCS/online_schema_change.md#L51)):</sup>
Looks
 like you need a forward/backward bit to specify the direction of the state change.


---

<sup>**[docs/RFCS/online_schema_change.md, line 152 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842#-K0fgd-jqUeAyY4hgQNQ-r3-152)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0574252634553b297160bc080ca5232f3af4e1/docs/RFCS/online_schema_change.md#L152)):</sup>
I quite convinced that what you are suggesting is fraught with complexity. In fact it wouldn't surprise me that you'd end up implementing transactions eventually. 


---

<sup>**[docs/RFCS/online_schema_change.md, line 162 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842#-K0fgd-jqUeAyY4hgQNQ-r3-162)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0574252634553b297160bc080ca5232f3af4e1/docs/RFCS/online_schema_change.md#L162)):</sup>
true, but it's ugly to change it and not update the version.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842)

**Comment by petermattis** [comment]
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

<sup>**[docs/RFCS/online_schema_change.md, line 51 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842#-K0fgd-jqUeAyY4hgQNQ-r3-51)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0574252634553b297160bc080ca5232f3af4e1/docs/RFCS/online_schema_change.md#L51)):</sup>
The forward/backward bit is captured in the `oneof mutation` info.


---

<sup>**[docs/RFCS/online_schema_change.md, line 152 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842#-K0fgd-jqUeAyY4hgQNQ-r3-152)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0574252634553b297160bc080ca5232f3af4e1/docs/RFCS/online_schema_change.md#L152)):</sup>
Definitely fraught with complexity. We should leave the non-transactional version of backfilling as a future optimization.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842)

**Comment by spencerkimball** [comment]
<sup>**[docs/RFCS/online_schema_change.md, line 152 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842#-K0fgd-jqUeAyY4hgQNQ-r3-152)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0574252634553b297160bc080ca5232f3af4e1/docs/RFCS/online_schema_change.md#L152)):</sup>
Actually, the tombstones for deletions are already part of the MVCC system. We currently don't create them if the key being deleted isn't actually present, but we certainly could allow them to be created. We could have the backfill generate write data with a historic timestamp equal to the timestamp of the schema change flip to "write only" and further, allow it to insert MVCC values at timestamps earlier than more recently written values. The backfill would proceed through the full primary index scan by reading at the same historic timestamp which data is being written.

In other words, the backfill would create exactly the data at that historic timestamp as it would if all mutations to the table were paused for the backfill to complete.

I'm quite sure (though couldn't find any mention in the paper) that F1 doesn't transactionally backfill. They make some mention of map-reduce...that doesn't sound compatible with transactions to me.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842)

**Comment by spencerkimball** [comment]
<sup>**[docs/RFCS/online_schema_change.md, line 152 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842#-K0fgd-jqUeAyY4hgQNQ-r3-152)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0574252634553b297160bc080ca5232f3af4e1/docs/RFCS/online_schema_change.md#L152)):</sup>
Look in http://static.googleusercontent.com/media/research.google.com/en//pubs/archive/41376.pdf, Section 4.4


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842)

**Comment by petermattis** [comment]
Review status: 0 of 1 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

<sup>**[docs/RFCS/online_schema_change.md, line 152 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842#-K0fgd-jqUeAyY4hgQNQ-r3-152)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0574252634553b297160bc080ca5232f3af4e1/docs/RFCS/online_schema_change.md#L152)):</sup>
@tschottdorf and I discussed this in the table_descriptor_lease RFC and it certainly looks possible to do the backfilling without transactions.  I agree it could be done, but I hope you'll agree that is more complex than the transactional approach described here and we should get that working first.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2842)

**Comment by spencerkimball** [comment]
No problem with starting transactionally as long as it's not too much work.
Let's also file an issue for adding the capability to mvcc to accommodate
naked delete tombstones and also a flag for allowing historical writes (or
simply ignoring them as I think F1 does). The issue should additionally
include a means to augment the write path through leader replicas such that
they do not affect the read time stamp cache.
